{% extends "base.html" %}

{% block head-extras %}
<title> CollecTF - transcription factor binding motif for
  {% for report in reports %}{{ report.TF_name }} in {{ report.species_name }}
  {% if not forloop.last %}, {% endif %}{% endfor %}
</title>
{% endblock %}

{% block content %}

{% load dbxref %}
{% load diagram %}

<div class="">
  {% if reports %}
  <div>
    <ul class="nav nav-pills">
      <li class="active"><a href="#split_view" data-toggle="tab">Split view</a></li>
      <li><a href="#ensemble_view" data-toggle="tab">Ensemble view</a></li>
      <li><a href="#export" data-toggle="tab">Export data</a></li>
    </ul>
  </div>

  <div class="tab-content">
    <div class="tab-pane active" id="split_view">
      {% for report in reports %}
      <div class="boxed">
	    <h3>{{ report.TF_name }}
          {% for acc in report.TF_accessions %}
          <a href="http://www.ncbi.nlm.nih.gov/protein/?term={{acc}}" target="_blank">
            <span class="fixed-width">{{ acc }}</span>
          </a>
          {% if not forloop.last %}, {% endif %}
          {% endfor %}
          binding site collection of 
          <i>{{ report.species_name }}</i>
        </h3>
        
	    <ul class="nav nav-tabs" id="sites_tab">
	      <li>
            <a href="#{{report.id}}_unaligned" data-toggle="tab">Binding sites</a>
          </li>
	      <li>
            <a href="#{{report.id}}_aligned" data-toggle="tab">Aligned binding sites</a>
          </li>
	      <li class="active">
            <a href="#{{report.id}}_weblogo" data-toggle="tab">Sequence logo</a>
          </li>
	      <li>
            <a href="#{{report.id}}_detail" data-toggle="tab">Detailed view</a>
          </li>
	    </ul>

	    <div class="tab-content">
	      <div class="tab-pane" id="{{report.id}}_unaligned">
	        <p>Sites are listed as curated.</p>
	        <pre>
	          {% for meta_site in report.meta_sites %}
              {{ meta_site.delegate.site_instance.seq }}{% endfor %}
	        </pre>
	      </div>
	      <div class="tab-pane" id="{{report.id}}_aligned">
	        <p>Sites are listed after the alignment process. For alignment of variable-length
	          binding sites, <a href="http://www.biomedcentral.com/1471-2105/14/108/"
	          target="_blank">LASAGNA</a> is used.
	        </p>
	        <pre>
	          {% for site in report.align_sites %}
              {{ site }}{% endfor %}
	        </pre>
	      </div>
	      <div class="tab-pane active" id="{{report.id}}_weblogo">
	        <p>To generate the weblogo, aligned binding sites are used.</p>
            <div class="">
              <img src="" class="weblogo_img"
                   site_data={% for site in report.align_sites %}{{ site }}{% if not forloop.last %},{% endif %}{% endfor %}>
            </div>

	      </div>
	      <div class="tab-pane" id="{{report.id}}_detail">
	        <p>For the selected transcription factor and species, the list of
	          curated binding sites in the database are displayed below. The
	          binding site is displayed as red bar. Genes that are reported to
	          be regulated by TF are colored with green. </p>
	        
            <ul id="hiddenCols"></ul>
	        <table class="table table-bordered table-striped">
	          <thead>
		        <tr>
		          <th class="genome">
                    Genome[<a href="#" class="" onclick="hideCol('genome');">x</a>]
                  </th>
		          <th class="TF">
                    TF[<a href="#" class="" onclick="hideCol('TF');">x</a>]
                  </th>
		          <th>Site sequence</th>
		          <th class="techniques">
                    Experimental techniques[<a href="#" onclick="hideCol('techniques');">x</a>]
                  </th>
		          <th class="diagram">
                    Gene regulation [<a href="" id="regulation_diagram_switch"><small>Toggle
                    diagrams</small></a>][<a href="#" onclick="hideCol('diagram');">x</a>]</th>
		          <th>Curations</th>
		        </tr>
	          </thead>

	          <tbody>
		        {% for meta_site in report.meta_sites %}
		        <tr {% if meta_site.site_type == 'var_motif_associated' %}
                    class="warning"
                    {% else %}
                    class=""
                    {% endif %}>

                  {% with genome_acc=meta_site.genome_accession %}
		          <td class="genome">
                    <a href="http://www.ncbi.nlm.nih.gov/nuccore/?term={{ genome_acc }}" target="_blank">
                      <span class="fixed-width">{{ genome_acc }}</span>
                    </a>
                  </td>
		          {% endwith %}
                  
		          {% with TF_accessions=meta_site.TF_instance_accessions%}
		          <td class="TF">
                    {% for TF_acc in TF_accessions %}
                    <a href="http://uniprot.org/uniprot/{{ TF_acc }}" target="_blank">
                      <span class="fixed-width">{{ TF_acc }}</span>
                    </a>
                    {% if not forloop.last %}, {% endif %}
                    {% endfor %}
                  </td>
                  {% endwith %}

		          <td>
                    <span class="sequence">{{ meta_site.delegate.site_instance.seq }}</span>
                  </a></td>
		          
		          <td class="techniques">
		            {% for technique in meta_site.techniques %}
                    {% include "experimental_technique.html" %}                    
                    {% if not forloop.last %} - {% endif %}
		            {% endfor %}
		          </td>
                  
		          <td class="diagram">
		            {% for reg in meta_site.regulations %}
		            <a href="http://www.ncbi.nlm.nih.gov/gene/{{ reg.gene.gene_accession }}"
                       target="_blank">
		              <span class="fixed-width
                                   {% if reg.evidence_type == 'exp_verified' %}
                                   exp_verified_reg
                                   {% endif %}">
                        {% with gene_name=reg.gene.name locus_tag=reg.gene.locus_tag %}
			            {{ gene_name }} {% if gene_name != locus_tag %}
                        ({{locus_tag }}) {% endif %}
                        {% endwith %}
                      </span>
                    </a>{% if not forloop.last %},{% endif %}
		            {% endfor %}

		            <div class="regulation_diagram">
		              {% with regulations=meta_site.regulations %}
		              {{ regulations|regulation_diagram}}
		              {% endwith %}
		            </div>
		          </td>

		          <td>{% for curation_id in meta_site.curation_ids %}
		            <span class="fixed-width">
		              <a href="{% url 'browse.view_curation.view_curation' curation_id %}"
                         target="_blank">
                        {{curation_id}}
                      </a>{% if not forloop.last %}, {% endif %}
		            </span>
		            {% endfor %}
		          </td>
		        </tr>
		        {% endfor %}

	          </tbody>
	        </table>
	      </div>
	    </div>
      </div>
      {% endfor %}
    </div>

    <div class="tab-pane" id="ensemble_view">
      <h3>Ensemble view</h3>
      
      <p>
        All binding sites in split view are combined and a sequence logo is generated. Note that it
	    may contain binding site sequences from different transcription factors and different
	    species. To see individiual sequence logos and curation details go to split view.
      </p>
      
      <div class="boxed">
	    <ul class="nav nav-tabs" id="sites_tab">
	      <li>
            <a href="#ensemble_view_unaligned" data-toggle="tab">Binding sites</a>
          </li>
	      <li>
            <a href="#ensemble_view_aligned" data-toggle="tab">Aligned binding sites</a>
          </li>
	      <li class="active">
            <a href="#ensemble_view_weblogo" data-toggle="tab">Sequence logo</a>
          </li>
	    </ul>

	    <div class="tab-content">
	      <div class="tab-pane" id="ensemble_view_unaligned">
	        <p>Sites are listed as curated.</p>
	        <pre>
	          {% for meta_site in ensemble_report.meta_sites %}
              {{ meta_site.delegate.site_instance.seq }}{% endfor %}
	        </pre>
	      </div>
	      <div class="tab-pane" id="ensemble_view_aligned">
	        <p>Sites are listed after the alignment process. For alignment of
	          variable-length binding
	          sites, <a href="http://www.biomedcentral.com/1471-2105/14/108/"
	          target="_blank">LASAGNA</a> is used.
	        </p>
	        <pre>
	          {% for site in ensemble_report.align_sites %}
              {{ site }}{% endfor %}
	        </pre>
	      </div>
	      <div class="tab-pane active" id="ensemble_view_weblogo">
	        <p>To generate the weblogo, aligned binding sites are used.</p>
	        <div class="">
	          <img src="" class="weblogo_img" id="{{report.id}}_weblogo"
                   site_data={% for site in ensemble_report.align_sites %}{{ site }}{% if not forloop.last %},{% endif %}{% endfor %}>
	        </div>
	      </div>
	    </div>
      </div>
    </div>
    
    <div class="tab-pane" id="export">
      <div class="boxed">
	    <form action="{% url 'browse.export.export_sites' %}" method="POST">
          {% csrf_token %}
          {% for report in reports %}
          {% for meta_site in report.meta_sites %}
	      <input type="hidden" name="site_id"
                 value="{% for csi_id in meta_site.curation_site_instance_ids %}
                        {{ csi_id }}{% if not forloop.last %}|{% endif %}{% endfor %}">
	      {% endfor %}
	      {% endfor %}

          <table class="table table-bordered table-striped">
            <tbody>
              <tr>
                <td>
                  <input type="submit" value="Download FASTA" class="btn btn-link" name="fasta">
                </td>
                <td>Download data in FASTA format.</td>
              </tr>
              <tr>
                <td><input type="submit" value="Download spreadsheet (TSV)"
                           class="btn btn-link" name="tsv">
                </td>
                <td>Download data in TSV (tab-separated-value) format. For each binding site, all
		          sources of evidence (i.e. experimental techniques and publication information) are
		          combined into one record.</td>
              </tr>
              <tr>
                <td>
                  <input type="submit" value="Download spreadsheet (TSV, raw)"
                         class="btn btn-link" name="tsv-raw">
                </td>
                <td>Download raw data in TSV format. All reported sites are exported
	              individually.</td>
              </tr>
	          <tr>
                <td><input type="submit" value="Download ARFF" class="btn btn-link" name="arff"></td>
                <td>Download data in Attribute-Relation File Format (ARFF).</td>
              </tr>
              <tr>
                <td>
                  <input type="submit" value="Download PSFM (TRANSFAC)"
                         class="btn btn-link" name="PSFM-transfac" />
                </td>
                <td>Download Position-Specific-Frequency-Matrix of the motif in TRANSFAC
                format.</td>
              </tr>
              <tr>
                <td>
                  <input type="submit" value="Download PSFM (JASPAR)"
                         class="btn btn-link" name="PSFM-jaspar" />
                </td>
                <td>Download Position-Specific-Frequency-Matrix of the motif in JASPAR format.</td>
              </tr>
              <tr>
                <td>
                  <input type="submit" value="Download PSFM (raw-FASTA)" class="btn btn-link"
                         name="PSFM-raw-fasta" />
                </td>
                <td>Download Position-Specific-Frequency-Matrix of the motif in raw FASTA
                  format. The matrix consists of four columns in the order A C G T.</td>
              </tr>
            </tbody>
          </table>
	    </form>

      </div>
    </div>
  </div>
  {% else %}
  <p>No results were found.</p>
  {% endif %}
</div>


<script type="text/javascript">
// TODO(sefa): move this to a .js file
function hideCol(colClass) {
    $('table .'+colClass).each(function() {
        $(this).toggle();
    });
    $('ul#hiddenCols').append('<li id="'+colClass
			                  +'"><a href="javascript:;" onclick="showCol(\''
			                  +colClass+'\');">Show '+colClass+ ' column</a></li>');
}

function showCol(columnClass){
    $('table .'+columnClass).each(function(index) {
        $(this).show();
    });
    $('li#'+columnClass).remove();
}

$(document).ready(function() {
    // regulation diagram hide/show
    $("#regulation_diagram_switch").on('click', function() {
        $(".regulation_diagram").toggle();
        return false;
    })

    $('.weblogo_img').each(function() {
        var img = $(this);
        img.parent().block({message: '<i class="icon-spinner icon-spin"></i>'});
        var site_data = $(this).attr('site_data');
        $.ajax({
            type: "POST",
            url: "{% url 'base.views.get_weblogo' %}",
            data: {
                sites: site_data,
                csrfmiddlewaretoken:'{{csrf_token}}', 
            },
            success: function(data) {
                img.attr('src', data);
                img.parent().unblock();
            },
        });
    });

    // hide 'genome' and 'TF' columns in the table
    hideCol('genome');
    hideCol('TF');

});
</script>

{% endblock %}
